rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,324 @@
+best of its ability, logging any corruption and repairs performed.
+
+To support the repair of the WAL, I propose introducing two new methods:
+`DB::wal_inspect()` and `DB::wal_repair()`. These methods will provide
+developers with tools to inspect the current contents of the WAL, including
+corrupted entries, and to repair the WAL to the best of its ability via entity
+recovery or truncation of the WAL.
+
+NOTE: `DbOptions.wal_auto_recovery` is a convenience for developers, as the
+same result can be achieved by calling `DB::open_with_opts()`, immediately
+calling `DB::wal_repair()`, and then calling `DB::open_with_opts()` again.
+
+### Handling SST Corruption
+Corruption of the SST can occur at any stage before, during, or after a
+database operation. Therefore, the most suitable place to report corruption is
+during the `get` call to retrieve keys from the database. If a key is NOT found
+and SlateDB detects corruption while searching for the key, an error should be
+reported. However, if SlateDB encounters corruption while searching for AND
+finding a key, it should not return an error. This ensures the database remains
+available even in the presence of corruption.
+
+The error returned should be a distinct type which can easily be matched to
+determine if the error is due to non transient corruption or a transient error.
+(network connectivity, etc..)
+
+The rationale behind this approach is rooted in the assumption that SlateDB's
+design is a manifestation of a physical system. Linux system administrators
+would not use physical disks that refuse access to all usable sectors due to a
+single bad sector. Such a disk would be impractical in many LVM (Logical Volume
+Manager) or RAID configurations. Similarly, the ability for developers who wish
+to use SlateDB to achieve Logical Durability is severely compromised if access
+to uncorrupted data blocks is not allowed.
+
+To support those operators who value durability, the developer making the call
+to `get` can decide to abort the application if and corruption error is returned.
+The application can then provide operators with detailed information included
+in the error returned by SlateDB on the nature of the corruption and suggestions
+on how to fix it.
+
+To support those operators who value availability, the error can be logged or
+ignored until either the database is deleted, or corruption repaired.
+
+To support repair of the SST, I propose we introduce new methods called
+`DB::sst_check()` , `DB::sst_inspect()` and `DB::sst_repair()` providing
+developers tools to validate the integrity of the entire database, inspect the
+current contents of a specific SST and repair a specific SST which has been
+identified as corrupt.
+
+> NOTE: We could optionally support verification of the entire database when
+> `DB::open_with_opts()` is called via some future `DBOptions` config. In
+> addition, we can expose the inspection and repair methods via a CLI available
+> to operators.
+
```

Comment:
The RFC should differentiate between corrupt SST file segments in disk cache and corrupt SST files or blocks I object storage (the source of truth). In the former, we should probably drop the file segments from disk and try to redownload/verify. In the latter, if the object store data is corrupt, then the previous sections all apply.